gateway-ha/src/test/java/io/trino/gateway/ha/TestPutRequestHandler.java

```diff
@@ -0,0 +1,122 @@
+import okhttp3.mockwebserver.Dispatcher;
+import okhttp3.mockwebserver.MockResponse;
+import okhttp3.mockwebserver.MockWebServer;
+import okhttp3.mockwebserver.RecordedRequest;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestInstance;
+import org.junit.jupiter.api.TestInstance.Lifecycle;
+
+import static com.google.common.net.HttpHeaders.CONTENT_TYPE;
+import static com.google.common.net.MediaType.JSON_UTF_8;
+import static org.assertj.core.api.Assertions.assertThat;
+
+@TestInstance(Lifecycle.PER_CLASS)
+public class TestPutRequestHandler
+{
+    private final OkHttpClient httpClient = new OkHttpClient();
+    private MockWebServer mockTrinoServer = new MockWebServer();
+
+    private final int routerPort = 21001 + (int) (Math.random() * 1000);
+    private final int customBackendPort = 21000 + (int) (Math.random() * 1000);
+
+    private static final String OK = "OK";
+    private static final int NOT_FOUND = 404;
+
+    private final String customPutEndpoint = "/v1/custom"; // this is enabled in test-config-template.yml
+    private final String healthCheckEndpoint = "/v1/info";
+
+    @BeforeAll
+    public void setup()
+            throws Exception
+    {
+        HaGatewayTestUtils.prepareMockBackend(mockTrinoServer, customBackendPort, "default custom response");
+        mockTrinoServer.setDispatcher(new Dispatcher() {
+            @Override
+            public MockResponse dispatch(RecordedRequest request)
+            {
+                if (request.getPath().equals(healthCheckEndpoint)) {
+                    return new MockResponse().setResponseCode(200)
+                            .setHeader(CONTENT_TYPE, JSON_UTF_8)
+                            .setBody("{\"starting\": false}");
+                }
+
+                if (request.getMethod().equals("PUT") && request.getPath().equals(customPutEndpoint)) {
+                    return new MockResponse().setResponseCode(200)
+                            .setHeader(CONTENT_TYPE, JSON_UTF_8)
+                            .setBody(OK);
+                }
+
+                return new MockResponse().setResponseCode(NOT_FOUND);
+            }
+        });
+
+        // seed database
+        HaGatewayTestUtils.TestConfig testConfig =
+                HaGatewayTestUtils.buildGatewayConfigAndSeedDb(routerPort, "test-config-template.yml");
+
+        // Start Gateway
+        String[] args = {testConfig.configFilePath()};
+        HaGatewayLauncher.main(args);
+
+        // Set up backend
+        HaGatewayTestUtils.setUpBackend(
+                "custom", "http://localhost:" + customBackendPort, "externalUrl", true, "adhoc", routerPort);
+    }
+
+    @AfterAll
+    public void cleanup()
+            throws Exception
+    {
+        if (mockTrinoServer != null) {
+            mockTrinoServer.shutdown();
+        }
+    }
+
+    @Test
+    public void testPutRequestHandler()
```

Comment:
Make test method package-private. https://trino.io/docs/current/develop/tests.html

Same for `setup` and `cleanup` methods.